File: dotnet/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example1/Singleton.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SingletonDesignPattern.Example1;

public class Singleton
{
	private static Singleton? _instance;

	private Singleton() { }

	public static Singleton GetOrCreateInstance()
	{
		if (_instance == null)
		{
			_instance = new Singleton();
		}

		return _instance;
	}
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example1/src/Singleton.ts
export class Singleton {
  private static _instance: Singleton | null = null;
  private static _idCounter: number = 0;
  public readonly id: number;

  private constructor() {
    Singleton._idCounter++;
    this.id = Singleton._idCounter;
  }

  static getOrCreateInstance(): Singleton {
    if (!Singleton._instance) {
      Singleton._instance = new Singleton();
    }
    return Singleton._instance;
  }
}
